### PR TITLE
Create "three-months roadmap" article

### PR DESCRIPTION
**Pull request by PoignardAzur**
[Rendered.](https://github.com/linebender/linebender.github.io/blob/e620332d675bd00af8f690e93a135b71c8a98f79/content/blog/2024-05-23-linebender-roadmap.md)

### CODE REVIEW

**Comment by DJMcNab** [comment]
[RENDERED](https://github.com/linebender/linebender.github.io/blob/roadmap/content/blog/2024-05-23-linebender-roadmap.md)

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+date = "2024-05-23 18:00:00"
++++
+
+On May 7 and 8, 2024, folks from the Linebender team went to [the RustNL conference](https://2024.rustnl.org/) for two days, followed by two days of Unconference.
+
+By that point, we all felt fairly exhausted, but Raph Levien being the cruel taskmaster that he is, he still made us work one extra day, and so we came together in our hotel's conference room to discuss the Linebender project, and what we wanted from it.
+
+We'd initially thought of that this day would be an informal day of unstructured chatting about the ecosystem, but it fairly quickly turned into a semi-formal planning exercise, where we tried to figure out a roadmap for the next few months.
+
+What follows is a comprehensive summary of what exactly we're planning to do. We've tried to focus on blockers, where a missing feature in on project is stalling progress on other project, because those are the hardest barriers to coordination, and we wanted to take advantage of being in the same room together to knock them down.
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:13]: One sentence per line please

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+- **Priority:**
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:15]: I'm not quite sure I follow what this section is doing here. Is the intention to fill out a brief description of these fields here?

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+- **Issue:** 
+
+
+## Xilem
+
+### Android bringup
+
+- **Priority:** Highest.
+- **Assigned:** Daniel McNab.
+- **Description:** We want Xilem to reliably work on Android. Running your Xilem app should be as simple as doing `cargo apk run` in your project.
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:155]: I'm not sure I want to be so specific about supporting exactly `cargo apk run`; there are concerns about how well it's possible to support Android with a non-gradle workflow. Additionally, we will almost certainly need at least some boilerplate in each app for Android support, which "in your project" *implies* wouldn't be needed.

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+- **PR:** [xilem#309](https://github.com/linebender/xilem/pull/309)
+
+### Update repository layout.
+
+- **Priority:** High.
+- **Assigned:** Olivier Faure.
+- **Description:** We've agreed on a layout for Linebender repository. We should move projects in the Xilem repo to match that layout.
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:162]:
```suggestion
- **Description:** We've agreed on a layout for Linebender repositories. We should move projects in the Xilem repo to match that layout.
```

File: content/blog/2024-05-23-linebender-roadmap.md (hunk 2 above)
Comment (PoignardAzur) [reply content/blog/2024-05-23-linebender-roadmap.md:15]: Oh yeah, that's a draft thing.

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+- **PR:** [vello#439](https://github.com/linebender/vello/pull/439)
+
+
+## Masonry
+
+### Complete text input
+
+- **Priority:** Highest.
+- **Assigned:** Olivier Faure, Aaron Muir Hamilton.
+- **Description:** We'd like to bring up Masonry's textbox widget to be best-in-class among Rust frameworks. We want to support native-feeling interactions out of the box, such as cursor movement (go to next word, go to beginning of line), selection (select to end of text) and edition (IME, deleting an entire word, etc). This is something other framework developers have expressed interest in, so we want to implement it in a composable way.
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:54]: One sentence per line please.
Also `edition` should probably be `editing`?

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+On May 7 and 8, 2024, folks from the Linebender team went to [the RustNL conference](https://2024.rustnl.org/) for two days, followed by two days of Unconference.
+
+By that point, we all felt fairly exhausted, but Raph Levien being the cruel taskmaster that he is, he still made us work one extra day, and so we came together in our hotel's conference room to discuss the Linebender project, and what we wanted from it.
+
+We'd initially thought of that this day would be an informal day of unstructured chatting about the ecosystem, but it fairly quickly turned into a semi-formal planning exercise, where we tried to figure out a roadmap for the next few months.
+
+What follows is a comprehensive summary of what exactly we're planning to do.
+We've tried to focus on blockers, where a missing feature in one project is stalling progress on other projects, because those are the hardest barriers to coordination, and we wanted to take advantage of being in the same room together to knock them down.
+
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:15]: We should probably have a sentence here that this document will not be updated with the current status, and so includes both planned and completed work.

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+- **Priority:** High.
+- **Assigned:** Olivier Faure.
+- **Description:** Masonry's documentation hasn't been updated since the port to Winit. While a lot of the code is still in flux, and thus hard to document, we should at least do a quick rewrite pass to remove references to Piet, Glazier, etc.
+- **PR:** [xilem#329](https://github.com/linebender/xilem/pull/329)
+
+### Finish repository port
+
+- **Priority:** High.
+- **Assigned:** Daniel McNab. (?)
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:74]:
```suggestion
- **Assigned:** Daniel McNab.
```
I've done this now, so it may as well be assigned to me less uncertainly

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+- **Priority:** Low.
+- **Assigned:** Olivier Faure.
+- **Description:** We need to figure out an architecture for doing unit tests in Xilem. Tests can then be added by volunteer contributors.
+- **Issue:** [xilem#393](https://github.com/linebender/xilem/issues/393)
+
+### Add basic benchmarks
+
+- **Priority:** Low.
+- **Assigned:** Olivier.
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:193]: We should be consistent in names here

File: content/blog/2024-05-23-linebender-roadmap.md (hunk 6 above)
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:15]: I think we should also have a comment about supporting community contribution both in these areas and beyond them (maybe? pointing to the new long term roadmap?)

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+- **Description:** Parley should support laying out inline content the way web browsers do, where blocks are splices into your text and should be laid out according to available inline space the same way words are.
+- **Issue:** [parley#25](https://github.com/linebender/parley/issues/25)
+- **PRs:** [parley#67](https://github.com/linebender/parley/pull/67), [parley#76](https://github.com/linebender/parley/pull/76)
+
+### Write adequate documentation
+
+- **Priority:** High.
+- **Assigned:** Nico Burns.
+- **Description:** Parley currently has virtually no documentation. If we want it to become a fundational crate, we need to explain to people how to use it.
+- **PR:** [parley#26](https://github.com/linebender/parley/pull/26)
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:136]: We really should try and just merge this "as-is"; it's not doing any good just sat as a draft.

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+- **Priority:** Medium.
+- **Assigned:** Daniel McNab.
+- **Description:** One of Xilem's selling point is the ability to handle virtual lists and async sources of data. That ability was temporarily removed for the Masonry port, but we'd like to add it back it. We don't want Masonry to be aware of async stuff, but the AppDriver trait should at least provide a method which lets Xilem access Masonry internals on its own initiative.
+- **Issue:** [xilem#340](https://github.com/linebender/xilem/issues/340)
+
+### Add unit tests
+
+- **Priority:** Low.
+- **Assigned:** Olivier Faure.
+- **Description:** We need to figure out an architecture for doing unit tests in Xilem. Tests can then be added by volunteer contributors.
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:187]: I'm not sure that we should "delegate" testing to external contributors; we should aim for it to be part of the process that everyone follows; and hopefully we can make it be very straightforward.

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+
+- **Priority:** Highest.
+- **Assigned:** Aaron Muir Hamilton, Raph Levien.
+- **Description:** The Java classes of an Android application are where the rubber meets the road. We're investigating what we want our boilerplate to look like.
+
+### Handling IME
+
+- **Priority:** Highest.
+- **Assigned:** Aaron Muir Hamilton, Raph Levien.
+- **Description:** We need to extend Winit to be compatible with Android IME interfaces. An Android application without IME is a non-starter. Our design choices here will have influence on accessibility as well.
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:226]: IME is very jargonny here; I think we should also mention virtual keyboards here, so that someone without domain knowledge can understand this item. I.e.
```suggestion
- **Description:** We need to extend Winit to be compatible with Android IME interfaces. An Android application without support for virtual keyboard and other input method is a non-starter. Our design choices here will have influence on accessibility as well.
```

File: content/blog/2024-05-23-linebender-roadmap.md [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+Whew! We have our work cut out for us there.
+
+Overall, I'm feeling very optimistic about the coming months.
+In the weeks since we first drafted this roadmap, we've already made a lot of progress on our respective tasks (for instance Daniel is now running on what async experts call a "work-stealing" algorithm), so it does look like we'll meet our deadlines.
+
+What comes after that is a bit hazier.
+Our three areas of focus are likely to be Android support, accessibility, and developer experience.
+
+As we fill out gaps in our feature set, Xilem is moving closer to becoming a stable, mature framework for writing GUI applications.
+We're not GUI yet, but we hope we'll be in the near future.
```
Comment (DJMcNab) [inline content/blog/2024-05-23-linebender-roadmap.md:246]: This is potentially "promising" more than we're willing to give (especially considering Raph's question answer in the RustNL talk). I like the optimistic sentiment.

I'd maybe rephrase it in terms of "filling out the features needed for Xilem to thrive"?

File: content/wiki/long-term-roadmap.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
++++
+title = "Post-July Linebender roadmap"
```
Comment (DJMcNab) [inline content/wiki/long-term-roadmap.md:2]: I think we need a title which emphasises that this is more of a living document, given the URL and position in the wiki

File: content/blog/2024-05-23-linebender-roadmap.md (hunk 6 above)
Comment (DJMcNab) [reply content/blog/2024-05-23-linebender-roadmap.md:15]: I see that this is partially reflected in the conclusion, but I still think it would also be useful to have a similar preface

**Review by DJMcNab — COMMENTED** [review]
Overall I think it will be really good to share this. A handful of details and potential additions, but nothing huge to change

Also, please move this to one sentence per line, including in the roadmap items; I believe bullet points in markdown properly support non-semantic newlines.

**Comment by PoignardAzur** [comment]
> Also, please move this to one sentence per line, including in the roadmap items; I believe bullet points in markdown properly support non-semantic newlines.

I really don't want to do that unless you think it's critical to stick to this standard.

Honestly, I'm skeptical of the one-sentence-per-line rule. So far the only result of that rule for me has been having to rewrite more stuff. I get that it's supposed to make diffs more readable for large paragraphs, but it also makes the document itself less readable and more annoying to write. When it comes to bullet points, writing them as one paragraph is natural and writing them one sentence per line needs even more fiddling.

It would be one thing if we had a canonical formatter that applied this rule, or if this was a living document that was going to see more diffs, but it's a blog post that will never be touched again after it's posted.

File: content/blog/2024-05-23-linebender-roadmap.md (hunk 6 above)
Comment (PoignardAzur) [reply content/blog/2024-05-23-linebender-roadmap.md:15]: I'll pass. I think the intro and conclusion are already too wordy.

**Comment by DJMcNab** [comment]
The one sentence per line thing is mostly for the sake of reviewing - not having it makes it harder to add specific comments about specific sentences, because GitHub doesn't let us comment on a range more fine-grained than a line. It means we have to do really awkward things like quoting a sentence within a paragraph to comment on it.
For the current state of this, it's fine as there is unlikely to be any more significant review, but I do think it makes reviewing text content much easier.

I'm not sure what we want the approval process for this to be; I think it's fine to land as-is, but maybe Raph will want to take a look?

**Comment by Philipp-M** [comment]
> but it also makes the document itself less readable and more annoying to write.

Isn't it rendered the same as without?

Btw. regarding rendering, to keep the link in the first post up-to-date (when the filename doesn't change at least), I'd suggest using something like:

https://github.com/linebender/linebender.github.io/blob/roadmap/content/blog/2024-05-23-roadmap-may-2024.md

(i.e. change the commit hash with the branch name)

File: content/blog/2024-05-23-roadmap-may-2024.md [hunk 14]
```diff
@@ -0,0 +1,250 @@
++++
+title = "May-July roadmap for Linebender"
+authors = ["Olivier Faure"]
++++
+
+On May 7 and 8, 2024, folks from the Linebender team went to [the RustNL conference](https://2024.rustnl.org/) for two days, followed by two days of Unconference.
+
+By that point, we all felt fairly exhausted, but Raph Levien being the cruel taskmaster that he is, he still made us work one extra day, and so we came together in our hotel's conference room to discuss the Linebender project, and what we wanted from it.
```
Comment (xStrom) [inline content/blog/2024-05-23-roadmap-may-2024.md:8]: I understand that this is a joke. However a surprising number of people failed to understand that [Raph's April 1st C++ post](https://raphlinus.github.io/rust/2023/04/01/rust-to-cpp.html) was a joke and that was a much more obvious joke. For that reason I recommend against using our blog for such banter, as too many people will take it literally.

```suggestion
By that point, we all felt fairly exhausted, but we still came together in our hotel's conference room to discuss the Linebender project, and what we wanted from it.
```

Comment (DJMcNab) [reply content/blog/2024-05-23-roadmap-may-2024.md:8]: I'm conflicted on this; I think it's probably right to remove it, but I wish it wasn't :laughing:

Comment (PoignardAzur) [reply content/blog/2024-05-23-roadmap-may-2024.md:8]: Yeah, that joke was funnier when I wrote it right after hell week. ^^

I'll soften it a bit, and make the joke framing more obvious.

Comment (PoignardAzur) [reply content/blog/2024-05-23-roadmap-may-2024.md:8]: (Also, that joke is a lot funnier when you know Raph and how laid back he is, but people reading a blog post probably won't know that, so the framing should be obvious.)

File: content/blog/2024-05-23-linebender-roadmap.md (hunk 12 above)
Comment (PoignardAzur) [reply content/blog/2024-05-23-linebender-roadmap.md]: I've changed the wording a bit to make it more of a "north star" thing.

**Comment by PoignardAzur** [comment]
I've fixed small errors and sanded down the more controversial/risky bits. I think this is ready to publish now.

File: content/wiki/long-term-roadmap.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+* Haven’t talked about it a lot
+  * Improve documentation (see [#402](https://github.com/AccessKit/accesskit/issues/402))
+  * Point to / create a repository of accessibility best practices.
+  * Not just code, also social: Have people with disabilities test your stuff!
+  * Point to resources, forums, etc? Need to talk to Matt about this. (What were the best practices at Microsoft?)
+  * Accessibility isn’t just screen readers. Also color contrast, customizing size, IME support, motor disabilities (avoid stuff that requires simultaneous input).
+
+* “Accessibility is more than adding a dependency to AccessKit”
+    * (Olivier: Note to self: include GMTK video about Celeste Assist Mode)
```
Comment (DJMcNab) [inline content/wiki/long-term-roadmap.md:61]: Does this need to be handled?

File: content/blog/2024-05-23-roadmap-may-2024.md
Comment (DJMcNab) [inline content/blog/2024-05-23-roadmap-may-2024.md:1]: File date probably needs to change

**Review by DJMcNab — APPROVED** [review]
Thanks! I think it's important not to let this slip farther, since we are (today) half way through the discussed period.

File: content/wiki/long-term-roadmap.md (hunk 15 above)
Comment (PoignardAzur) [reply content/wiki/long-term-roadmap.md:61]: Maybe if I get around to writing an article about accessibility. In the meantime, those are unstructured notes.

**Comment by PoignardAzur** [comment]
Will publish this on monay (and change the date then), to leave some breathing room between the two articles.
